docarray/array/storage/redis/backend.py

```diff
@@ -0,0 +1,188 @@
+
+import numpy as np
+from docarray import Document
+from docarray.array.storage.base.backend import BaseBackendMixin, TypeMap
+from docarray.helper import dataclass_from_dict
+
+from redis import Redis
+from redis.commands.search.field import NumericField, TextField, VectorField
+from redis.commands.search.indexDefinition import IndexDefinition
+
+if TYPE_CHECKING:
+    from ....typing import ArrayType, DocumentArraySourceType
+
+
+@dataclass
+class RedisConfig:
+    n_dim: int
+    host: str = field(default='localhost')
+    port: int = field(default=6379)
+    index_name: str = field(default='idx')
+    flush: bool = field(default=False)
+    update_schema: bool = field(default=True)
+    distance: str = field(default='COSINE')
+    redis_config: Dict[str, Any] = field(default_factory=dict)
+    index_text: bool = field(default=False)
+    tag_indices: List[str] = field(default_factory=list)
+    batch_size: int = field(default=64)
+    method: str = field(default='HNSW')
+    ef_construction: int = field(default=200)
+    m: int = field(default=16)
+    ef_runtime: int = field(default=10)
+    initial_cap: Optional[int] = None
+    block_size: Optional[int] = None
+    columns: Optional[List[Tuple[str, str]]] = None
+
+
+class BackendMixin(BaseBackendMixin):
+    """Provide necessary functions to enable this storage backend."""
+
+    TYPE_MAP = {
+        'str': TypeMap(type='text', converter=TextField),
+        'bytes': TypeMap(type='text', converter=TextField),
+        'int': TypeMap(type='integer', converter=NumericField),
+        'float': TypeMap(type='float', converter=NumericField),
+        'double': TypeMap(type='double', converter=NumericField),
+        'long': TypeMap(type='long', converter=NumericField),
+        'bool': TypeMap(type='long', converter=NumericField),
+    }
+
+    def _init_storage(
+        self,
+        _docs: Optional['DocumentArraySourceType'] = None,
+        config: Optional[Union[RedisConfig, Dict]] = None,
+        **kwargs,
+    ):
+        if not config:
+            raise ValueError('Empty config is not allowed for Redis storage')
+        elif isinstance(config, dict):
+            config = dataclass_from_dict(RedisConfig, config)
+
+        if config.distance not in ['L2', 'IP', 'COSINE']:
+            raise ValueError(
+                f'Expecting distance metric one of COSINE, L2 OR IP, got {config.distance} instead'
+            )
+        if config.method not in ['HNSW', 'FLAT']:
+            raise ValueError(
+                f'Expecting search method one of HNSW OR FLAT, got {config.method} instead'
+            )
+
+        if config.redis_config.get('decode_responses'):
+            config.redis_config['decode_responses'] = False
+
+        self._offset2id_key = 'offset2id__' + config.index_name
+        self._config = config
+        self.n_dim = self._config.n_dim
+        self._doc_prefix = "doc__" + config.index_name + ":"
```

Comment:
will we remove this `"doc__"` as discussed in the call? Also, please use `'` instead of `"`,  just to maintain a consistent style.